### ISSUE TITLE
256x20 network results

### ISSUE THREAD

**Issue by zediir** [leela-zero/leela-zero#884]
Here is the 256x20 network I trained. https://drive.google.com/open?id=1PSyfsDvXmtIrhUx4Gf6x7mGzDtwe4npr

I ran it on CGOS with the following settings

```text
%section player
     name      LZH256x20-t4-p1600
     invoke    leelaz.exe -w leelaz-model-3008000.txt -r1 --noponder -p 1600 -t4 -g --gpu 0 --gpu 1

%section player
     name      LZH256x20-t4-p3200
     invoke    leelaz.exe -w leelaz-model-3008000.txt -r1 --noponder -p 3200 -t4 -g --gpu 0 --gpu 1

%section player
     name      LZH256x20-t4-nolim
     invoke    leelaz.exe -w leelaz-model-3008000.txt -r1 -t4 -g --gpu 0 --gpu 1
```

The hardware it ran on is an i7 7700k and two GTX 1080 graphics cards.

The bayes elos were:
LZH256x20-t4-p1600: 3535
LZH256x20-t4-p3200: 3586
LZH256x20-t4-nolim: 3659

I also ran it against Leela 0.11 with equal time and hardware on kgs

>engine=leelaz.exe -w leelaz-model-3008000.txt -r10 -t2 -g --gpu 0 --noponder
>name=TygemLeela
>rules.time=5:00+10x0:15

>engine=Leela0110GTP_OpenCL.exe -t4 -g --noponder
>name=LeeaEleven

The result currently is:
TygemLeela: 27
LeeaEleven: 9

Gets caught in ladders sometimes and sometimes has problems with life and death.

Leelaz binary used was built from current next-branch.

https://www.gokgs.com/gameArchives.jsp?user=LeeaEleven

<details>
  <summary>Code used to get wins numbers from kgs</summary>
  
```javascript
var nodes = window.document.getElementsByClassName('grid')[0].childNodes
var results = {}
for(childNodeIndex in nodes) {
  var tds = nodes[childNodeIndex].childNodes;
  if(nodes[childNodeIndex].localName === "tr" && tds[1].localName === "td") {
    var white = tds[1].innerText
    var black = tds[2].innerText
    if(results[black] === undefined) {
      results[black] = 0
    }
    if(results[white] === undefined) {
      results[white] = 0
    }
    var result = tds[6].innerText[0]
    if(result == "B") {
      results[black]++
    }
    if(result == "W") {
      results[white]++
    }
  }
}
for(resultIndex in results) {
  console.log(resultIndex.substr(0,resultIndex.indexOf("[") - 1) + ': ' + results[resultIndex])
}
```
</details>

**Comment by l1t1** [comment]
so we now have bron many robots from leelaz. nice job.

**Comment by RavnaBergsndot** [comment]
Is this leelaz an FPU-tuned one?

**Comment by jkiliani** [comment]
Must be FPU-tuned, or @zediir would not have mentioned

> Leelaz binary used was built from current next-branch.

**Comment by tterava** [comment]
Did you train this from self-play data completely? Looks very promising!

**Comment by WhiteHalmos** [comment]
This is supervised, not self-play

**Comment by zediir** [comment]
Yes. From the TYGEM dataset from here https://github.com/yenw/computer-go-dataset

**Comment by jillybob** [comment]
There was someone who had an updated dataset, I beleive they pulled many many games from FOX. I can't find the dataset but it had 9Ds playing. It would be good if someone could find it, and provide a link to it to zediir to see if this other dataset would produce a stronger LeelaZero

**Comment by CamWagner** [comment]
@jillybob that's my dataset. im probably going to continue to update it with all current Tygem 9d games. currently it has fewer 9d games than the Tygem dataset.

**Comment by FFLaguna** [comment]
@zediir How many steps and what batch size did you train this network for? What was your final learning rate and momentum? And what were your Tensorboard Accuracy, MSE error, and Policy error? I'm extremely interested in comparing with your results.

**Comment by zediir** [comment]
3,008,000 steps with batch size 64. Accuracy was 59%. For the rest I don't remember and I'm done with this for now. Going to be gaming again after having my gpu busy for a month.

**Comment by gcp** [comment]
Is this the strongest network we think we have?

I would put it up as a reference on the graph, perhaps?

**Comment by jkiliani** [comment]
actually, why not 99e274c5?

**Comment by gcp** [comment]
Even better, yes.

**Comment by gcp** [comment]
Hmm, even gzipped it's a 92M download. I can't actually upload it without changing the server config.

**Comment by zediir** [comment]
Maybe someone could just run the match and then add the result manually to the database. Unfortunately my gpu usage is spoken for for a couple of more weeks before I can contribute again.

**Comment by gcp** [comment]
I actually killed the server trying to upload it. node unzips a copy (or even more than 1) in memory and the machine is swapping.

It should recover in a few minutes. I hope.

**Comment by gcp** [comment]
Feeling pretty stupid right now. Lesson learned: we're not quite ready for a 256x20 run.

**Comment by jkiliani** [comment]
Shit happens. What is the actual capacity limit of files in the server anyway? It was still able to handle the 128x10 nets after all...

**Comment by zediir** [comment]
At least you didn't write `chown www-data:www-data -R /var` when you meant to write `chown -R www-data:www-data var/` . (that might have happened to someone today)

**Comment by gcp** [comment]
Not sure what the limit on the server is. It's running mongodb and node, which are the big memory eaters. The file is "only" 270M but I suspect it balloons up a lot inside node as that isn't exactly known for being very memory efficient.

**Comment by jkiliani** [comment]
zero-test.sjeng.org is a backup server, right? If it has the same hardware characteristics, could it be used to safely experiment on this?

**Comment by zediir** [comment]
If I'm reading this right the only reason the server unzips the file is to calculate the hash for it?

**Comment by FFLaguna** [comment]
@gcp Very minor issue:  leelaz.exe will overwrite previous tuning files for different block/filter counts. So if somebody did a full tuning, it will be overwritten by a quick tuning on 20x256 when they run the weights the first time. When they go back to 6x128 it will overwrite that again with another quick tuning.

In effect, people might not realize their full tunings got reset to quick tunings. I recommend modifying leelaz.exe to rename the tuning files to include the block and filter count in the filename, to keep this separate. I have been doing so manually with my own custom builds, but I don't know how to get this to happen automatically.

**Comment by Splee99** [comment]
leelaz-model-3008000 is very strong. On my computer it won a even game with AQ in just 123 moves, 12 sec per move.

**Comment by Mezoka** [comment]
20 blocks still cannot handle large group properly, it gives over 90% win rate for black at the end of these games :  [Game 1](http://eidogo.com/#BNG2VSK4)  [Game 2](http://eidogo.com/#CBE6JCmW)

**Comment by jkiliani** [comment]
This 20 block net is pure supervised learning from professional games. I think it's highly likely that a RL 20 block net will do fine with large groups in general.

**Comment by Eddh** [comment]
I'm sure even 20 blocks will still fail with very large groups sometimes. I think this is what pushed deepmind to go for 40 blocks.

**Comment by jkiliani** [comment]
It may fail sometimes, but I'm pretty sure it will fail far less than the SL network. I think the Leela Zero net trained by @zediir may well be comparable to the Deepmind supervised learning reference net they show in the Alphago Zero paper in Fig. 3, which ended up around 200 Elo below Alphago Lee. The reinforcement learning net on (presumably) the same network architecture 256x20 was ~800 Elo stronger than this SL net.

**Comment by billyswong** [comment]
@jkiliani In this sense a 256x20 network trained by LeelaZero existing self-play games may be overall more balanced in strength and made less mistakes?

**Comment by CamWagner** [comment]
self play should be able to work out holes in its play right now

**Comment by jkiliani** [comment]
@billyswong I honestly have no idea. In theory, the quality of play of the professional game dataset ist much higher than current LZ self-play games, but network training can work in mysterious ways. Someone might try this, but training such a network is a long investment...

I do hope we end up testing this net on the server at some point, as another reference point better than best_v1.

**Comment by zediir** [comment]
There are two problems with training on high rank human games.

1. A lot of the games end in resignation so less endgame
2. No examples of bad moves to see why not to play them

**Comment by tapsika** [comment]
> I'm sure even 20 blocks will still fail with very large groups sometimes. I think this is what pushed deepmind to go for 40 blocks.

For 20 blocks I don't think a large group is so different than a smaller one, 40 blocks was probably tried simply because DM needed to surpass Master for marketing reasons (they are sensitive to such cosmetics, up to the levels of manipulating their results a bit, like in the AZ vs. AGZ case).

**Comment by Eddh** [comment]
There is already evidence that even 20 blocks has issues with large groups here #708. Granted this is a supervised network and not a reinforced one but still, there is evidence that even 20 blocks might have problems, no evidence that it will be perfect. I agree that the marketing effect of surpassing master played a part in the decision though.

**Comment by tapsika** [comment]
Since 20 blocks should have no noticeable range limit, policy errors are likely caused by other things. In those cases, it's not clear if 40 blocks would do better with the same training data and method. It allows for more complex computations so has a general advantage, but in a particular error case better / more training may have more effect.

Or maybe - just a wild idea - some things are more useful when can travel distances along the group in a back-and-forth manner (so 20 blocks can fall just a few layers short sometimes)? Sounds unlikely though, at least without filter starvation.

**Comment by gcp** [comment]
>zero-test.sjeng.org is a backup server, right?

It's the same physical machine.

**Comment by hydrogenpi** [comment]
So is this model (30080) stronger than original Leela 11 on a single GTX 1080/Ti?

**Comment by sethtroisi** [comment]
closing old issue
